File: Homeworks/Chapter1/Solutions/Soal6.cpp
#include <iostream>
#include <string>
using namespace std;
class node{
	friend class Doubly_Linked_List;
	private:
		int data;
		node *next;
		node *prev;
	public:
		node(){

		}
		node(int data,node *next=nullptr,node *prev=nullptr){
			this->data = data;
			this->next = next;
			this->prev = prev;
		}	
};

class Doubly_Linked_List{

	private:
		int size=0;
		node *head = nullptr;
		node *tail = nullptr;
	public:
		Doubly_Linked_List(){

		}
		Doubly_Linked_List(int n){
			this->size = n;
			if(n>0){
			node *s = new node (1);
			this->head = s;
			this->tail = s;
			node *p = s;
			for(int i=2;i<=n;i++){
				node * temp = new node (i,nullptr,p);
				p->next = temp;
				p = temp;
				}
				this->tail = p;
			}
			
		}
		void rev(int i,int j){
			if(i!=j&&this->size!=0&&j<=this->size){
				node *f;
				node *e;
				node *tmpi  = this->head;
				node *tmpj;
				node *cur;
				node *nx;
				for(int k=1;k<i;k++)
					tmpi = tmpi->next;
				
				tmpj=tmpi;
				for(int k=i;k<j;k++){
					tmpj = tmpj->next;
				}
			 	f = tmpi->prev;
				e = tmpj->next;
				if(i==1)this->head = tmpj;
				else if(tmpi->prev!=nullptr){
					(tmpi->prev)->next = tmpj;
				}
				if(j==this->size)this->tail = tmpi;
				else if(tmpj->next!=nullptr){
					(tmpj->next)->prev = tmpi;
				}
				cur=tmpi->next;
				nx = tmpi->next->next;
				tmpi->prev = tmpi->next;
				tmpi->next = e;
				while(cur!=tmpj&&cur!=nullptr){

				  	cur->next = cur->prev;
				 	cur->prev = nx;
				 	cur = nx;
				  	if(nx!=nullptr)nx = nx->next;
				  	else break;
				}
				tmpj->next = tmpj->prev;
				tmpj->prev = f;

			}
	}
		void print(){
			node *tmp = this->head;
			while(tmp!=nullptr){

				cout<<tmp->data<<' ';
				tmp=tmp->next;
			}
			cout<<endl;
		}
		void name(int k){
			if(this->size!=0&&k<=this->size){
				node* tmp=this->head;
				for(int i=1;i<k;i++){
					tmp=tmp->next;
				}
				cout<<tmp->data<<endl;
			}
		}
		~Doubly_Linked_List(){
			node *tmp = this->head;
			node *tmpnx;
			while(tmp!=nullptr){
				tmpnx = tmp->next;
				delete tmp;
				tmp = tmpnx;

			}
			this->head = nullptr;
			this->tail = nullptr;
			this->size = 0;
		}

};



int main(){
	int n,t;
	cin >> n >> t;
	Doubly_Linked_List  cassettes (n);
	while(t--){
		string cm;
		cin >> cm;
		if(cm=="NAME"){
			int k;
			cin >> k;
			cassettes.name(k);
		}
		if(cm=="REV"){
			int i,j;
			cin >> i >> j;
			cassettes.rev(i,j);
		}
	}
	return 0;

}

File: Homeworks/Chapter1/Solutions/Soal6.java
package DS_Chapter1;

import java.util.Scanner;

public class Soal6
{
    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        DoublyLinkedList list = new DoublyLinkedList();
        for (int i = 1 ; i<=n ; i++)
        {
            list.add(i);
        }
        int m = sc.nextInt();
        for (int j =0 ; j<m ; j++)
        {
            String command = sc.next();
            if(command.equals("NAME"))
            {
                System.out.println(list.get(sc.nextInt()));
            }
            else
            {
                int l = sc.nextInt() ;
                int h = sc.nextInt() ;
                list.reverse(l, h);
            }
        }

    }
}
class Node
{
    int data ;
    Node next ;
    Node prev ;
    public Node (int x  , Node prev , Node next)
    {
        this.data = x ;
        this.next = next ;
        this.prev = prev ;
    }
}
class DoublyLinkedList
{
    Node head ;
    Node tail ;
    public DoublyLinkedList()
    {
    }
    public void add(int x)
    {
        Node newNode = new Node(x , tail , null);
        if (head == null)
        {
            head = tail = newNode ;
        }
        else
        {
            tail.next = newNode ;
            tail = newNode ;
        }
    }
    private Node getNode(int index)
    {
        int k = 1 ;
        Node t = head ;
        while(index>k)
        {
            if(t.next != null)
                t = t.next ;
            else
                throw new IndexOutOfBoundsException();
            k++ ;
        }
        return t ;

    }
    public int get(int index)
    {
        return getNode(index).data ;
    }
    public void reverse(int l , int h)
    {
        Node start = getNode(l);
        Node end = getNode(h);
        while (l<h)
        {
            int temp = start.data ;
            start.data = end.data ;
            end.data = temp ;
            l++ ;
            h-- ;
            start = start.next ;
            end = end.prev ;
        }
    }
}